jxline-core/src/main/java/cloud/xline/jxline/impl/ProtocolClientImpl.java

```diff
@@ -0,0 +1,406 @@
+                if (ex.shouldAbortFastRound()) {
+                    throw ex;
+                }
+                if (exception == null || exception.priority().value() <= ex.priority().value()) {
+                    exception = ex;
+                }
+            }
+        }
+        if (exception != null) {
+            throw exception;
+        }
+        // We will at least send the request to the leader if no `WrongClusterVersion` returned.
+        // If no errors occur, the leader should return the ER
+        // If it is because the super quorum has not been reached, an error will definitely occur.
+        // Otherwise, there is no leader in the cluster state currently, return wrong cluster
+        // version
+        // and attempt to retrieve the cluster state again.
+        throw new CurpException(
+                CurpError.newBuilder().setWrongClusterVersion(Empty.newBuilder().build()).build());
+    }
+
+    <T> CompletionService<T> forEachServer(
+            Collection<VertxProtocolGrpc.ProtocolVertxStub> stubs,
+            Invoke<VertxProtocolGrpc.ProtocolVertxStub, T> task) {
+        CompletionService<T> completionService =
+                new ExecutorCompletionService<>(this.connectionManager().getExecutorService());
+        for (VertxProtocolGrpc.ProtocolVertxStub stub : stubs) {
+            completionService.submit(() -> task.call(stub));
+        }
+        return completionService;
+    }
+
+    <T> Future<T> mapLeader(Function<VertxProtocolGrpc.ProtocolVertxStub, Future<T>> task) {
+        VertxProtocolGrpc.ProtocolVertxStub leaderStub = this.state.getLeaderStub();
+        if (leaderStub == null) {
+            // choose a random leader, it will return redirect error if leader is wrong
+            // TODO: fetch cluster here.
+            leaderStub =
+                    this.state.stubs.values()
+                            .toArray(VertxProtocolGrpc.ProtocolVertxStub[]::new)[0];
+        }
+        return task.apply(leaderStub);
+    }
+
+    int superQuorum(int size) {
+        int faultTolerance = size - quorum(size);
+        return faultTolerance + recoverQuorum(size);
+    }
+
+    int quorum(int size) {
+        return size / 2 + 1;
+    }
+
+    int recoverQuorum(int size) {
+        return quorum(size) / 2 + 1;
+    }
+
+    State getInitState() {
+        ManagedChannel initChannel = this.connectionManager().getInitChannel();
+        VertxProtocolGrpc.ProtocolVertxStub initStub = VertxProtocolGrpc.newVertxStub(initChannel);
+        FetchClusterResponse response = null;
+        FetchClusterRequest request =
+                FetchClusterRequest.newBuilder().setLinearizable(false).build();
+        int retries = -1;
+        do {
+            retries++;
+            if (retries > 5) {
+                throw new RuntimeException("connection failed");
+            }
+            try {
+                response = completable(initStub.fetchCluster(request)).get(3, TimeUnit.SECONDS);
```

Comment:
please declare constants instead of using magic numbers.

Reply:
This method should be refactored in the future(i.e. implementing the retry policy). I'll leave a TODO here.

Reply:
So at least create a const named `retry_times_limit` instead of using 3 directly.

Reply:
done